File: packages/protocol/src/registry.ts
import registryJson from '../registry.json';

/**
 * The registry is the explicit allowlist of everything the renderer knows how
 * to display. Nothing outside of it can ever reach the renderer: the event
 * server rejects unknown ids and the renderer ignores them defensively.
 */

export const SCENE_IDS = ['city', 'forest', 'space', 'roadside-workshop'] as const;
export const EFFECT_IDS = ['rain', 'snow', 'fireworks'] as const;
export const EVENT_SOURCES = ['manual', 'simulation', 'system'] as const;
/**
 * Scene actions: named, predefined things a scene can be asked to do once,
 * such as sending a bus down the street. An action id selects a capability the
 * renderer already has; it never describes behaviour.
 */
export const SCENE_ACTION_IDS = [
  'roadside.send-car',
  'roadside.send-bus',
  'roadside.pedestrians',
  'roadside.blow-leaves',
  'roadside.rush-hour',
  'forest.bird-flock',
  'space.shooting-star',
] as const;

export type SceneId = (typeof SCENE_IDS)[number];
export type EffectId = (typeof EFFECT_IDS)[number];
export type EventSource = (typeof EVENT_SOURCES)[number];
export type SceneActionId = (typeof SCENE_ACTION_IDS)[number];

export interface RegistryEntry<Id extends string> {
  readonly id: Id;
  readonly label: string;
  readonly description: string;
}

export type SceneEntry = RegistryEntry<SceneId>;

export interface EffectEntry extends RegistryEntry<EffectId> {
  /** `null` means the effect runs until it is explicitly cleared. */
  readonly defaultDurationMs: number | null;
}

export interface SceneActionEntry extends RegistryEntry<SceneActionId> {
  /** The only scene this action applies to. */
  readonly sceneId: SceneId;
  /** Minimum time between two accepted requests for this action. */
  readonly cooldownMs: number;
}

export interface Registry {
  readonly protocolVersion: number;
  readonly scenes: readonly SceneEntry[];
  readonly effects: readonly EffectEntry[];
  readonly actions: readonly SceneActionEntry[];
  readonly sources: readonly EventSource[];
}

export const REGISTRY = registryJson as Registry;

export const PROTOCOL_VERSION = REGISTRY.protocolVersion;

export const DEFAULT_SCENE_ID: SceneId = 'city';

/** Bounds on a registry action cooldown, shared with the event server. */
export const MIN_ACTION_COOLDOWN_MS = 250;
export const MAX_ACTION_COOLDOWN_MS = 60_000;

function assertIdsMatch(declared: readonly string[], fromJson: readonly string[], what: string): void {
  const a = [...declared].sort().join(',');
  const b = [...fromJson].sort().join(',');
  if (a !== b) {
    throw new Error(`registry.json drifted from the declared ${what} union: [${b}] vs [${a}]`);
  }
}

// Fail loudly at import time rather than silently serving ids the types deny.
assertIdsMatch(
  SCENE_IDS,
  REGISTRY.scenes.map((scene) => scene.id),
  'scene id',
);
assertIdsMatch(
  EFFECT_IDS,
  REGISTRY.effects.map((effect) => effect.id),
  'effect id',
);
assertIdsMatch(
  SCENE_ACTION_IDS,
  REGISTRY.actions.map((action) => action.id),
  'scene action id',
);
assertIdsMatch(EVENT_SOURCES, REGISTRY.sources, 'event source');
for (const action of REGISTRY.actions) {
  if (!(SCENE_IDS as readonly string[]).includes(action.sceneId)) {
    throw new Error(`registry.json action ${action.id} names an unknown scene: ${action.sceneId}`);
  }
  if (
    !Number.isInteger(action.cooldownMs) ||
    action.cooldownMs < MIN_ACTION_COOLDOWN_MS ||
    action.cooldownMs > MAX_ACTION_COOLDOWN_MS
  ) {
    throw new Error(`registry.json action ${action.id} has an out-of-range cooldownMs`);
  }
}

export function isSceneId(value: unknown): value is SceneId {
  return typeof value === 'string' && (SCENE_IDS as readonly string[]).includes(value);
}

export function isEffectId(value: unknown): value is EffectId {
  return typeof value === 'string' && (EFFECT_IDS as readonly string[]).includes(value);
}

export function isEventSource(value: unknown): value is EventSource {
  return typeof value === 'string' && (EVENT_SOURCES as readonly string[]).includes(value);
}

export function isSceneActionId(value: unknown): value is SceneActionId {
  return typeof value === 'string' && (SCENE_ACTION_IDS as readonly string[]).includes(value);
}

export function sceneEntry(id: SceneId): SceneEntry {
  const entry = REGISTRY.scenes.find((scene) => scene.id === id);
  if (!entry) throw new Error(`unknown scene id: ${id}`);
  return entry;
}

export function effectEntry(id: EffectId): EffectEntry {
  const entry = REGISTRY.effects.find((effect) => effect.id === id);
  if (!entry) throw new Error(`unknown effect id: ${id}`);
  return entry;
}

export function sceneActionEntry(id: SceneActionId): SceneActionEntry {
  const entry = REGISTRY.actions.find((action) => action.id === id);
  if (!entry) throw new Error(`unknown scene action id: ${id}`);
  return entry;
}

/** The actions a scene owns, in registry order. */
export function actionsForScene(sceneId: SceneId): readonly SceneActionEntry[] {
  return REGISTRY.actions.filter((action) => action.sceneId === sceneId);
}

File: services/event-server/src/livescape_event_server/registry.py
"""Explicit allowlist of scenes, effects and scene actions the renderer supports.

This mirrors ``packages/protocol/registry.json``; ``tests/test_registry.py``
fails if the two copies drift apart.
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any, Literal, get_args

REGISTRY_PATH = Path(__file__).with_name("registry.json")

SceneId = Literal["city", "forest", "space", "roadside-workshop"]
EffectId = Literal["rain", "snow", "fireworks"]
EventSource = Literal["manual", "simulation", "system"]
SceneActionId = Literal[
    "roadside.send-car",
    "roadside.send-bus",
    "roadside.pedestrians",
    "roadside.blow-leaves",
    "roadside.rush-hour",
    "forest.bird-flock",
    "space.shooting-star",
]

SCENE_IDS: tuple[str, ...] = get_args(SceneId)
EFFECT_IDS: tuple[str, ...] = get_args(EffectId)
EVENT_SOURCES: tuple[str, ...] = get_args(EventSource)
SCENE_ACTION_IDS: tuple[str, ...] = get_args(SceneActionId)

PROTOCOL_VERSION = 1

#: Bounds on a registry action cooldown, shared with ``registry.ts``.
MIN_ACTION_COOLDOWN_MS = 250
MAX_ACTION_COOLDOWN_MS = 60_000


@lru_cache(maxsize=1)
def load_registry() -> dict[str, Any]:
    """Load and sanity-check the registry document."""
    data: dict[str, Any] = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))

    declared_scenes = tuple(entry["id"] for entry in data["scenes"])
    declared_effects = tuple(entry["id"] for entry in data["effects"])
    if sorted(declared_scenes) != sorted(SCENE_IDS):
        raise RuntimeError(f"registry.json scene ids {declared_scenes} != {SCENE_IDS}")
    if sorted(declared_effects) != sorted(EFFECT_IDS):
        raise RuntimeError(f"registry.json effect ids {declared_effects} != {EFFECT_IDS}")
    declared_actions = tuple(entry["id"] for entry in data["actions"])
    if sorted(declared_actions) != sorted(SCENE_ACTION_IDS):
        raise RuntimeError(f"registry.json action ids {declared_actions} != {SCENE_ACTION_IDS}")
    for entry in data["actions"]:
        if entry["sceneId"] not in SCENE_IDS:
            raise RuntimeError(f"registry.json action {entry['id']} names an unknown scene")
        cooldown = entry["cooldownMs"]
        if (
            not isinstance(cooldown, int)
            or not MIN_ACTION_COOLDOWN_MS <= cooldown <= MAX_ACTION_COOLDOWN_MS
        ):
            raise RuntimeError(f"registry.json action {entry['id']} has an out-of-range cooldown")
    if sorted(data["sources"]) != sorted(EVENT_SOURCES):
        raise RuntimeError(f"registry.json sources {data['sources']} != {EVENT_SOURCES}")
    if data["protocolVersion"] != PROTOCOL_VERSION:
        raise RuntimeError("registry.json protocolVersion does not match the server")
    return data


def default_duration_ms(effect_id: str) -> int | None:
    """Duration an effect runs for when the event does not specify one."""
    for entry in load_registry()["effects"]:
        if entry["id"] == effect_id:
            duration: int | None = entry["defaultDurationMs"]
            return duration
    raise KeyError(f"unknown effect id: {effect_id}")


def action_entry(action_id: str) -> dict[str, Any]:
    """The registry entry of a scene action: owning scene, label and cooldown."""
    for entry in load_registry()["actions"]:
        if entry["id"] == action_id:
            return entry  # type: ignore[no-any-return]
    raise KeyError(f"unknown scene action id: {action_id}")
